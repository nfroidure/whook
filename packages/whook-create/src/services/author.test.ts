/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, test, beforeEach, jest, expect } from '@jest/globals';
import _inquirer from 'inquirer';
import initAuthor from './author.js';
import { YError } from 'yerror';
import { type LogService, type LockService } from 'common-services';

describe('initAuthor', () => {
  const exec = jest.fn() as any;
  const lock = {
    take: jest.fn<LockService<unknown>['take']>(),
    release: jest.fn<LockService<unknown>['release']>(),
  };
  const inquirer = { prompt: jest.fn<(typeof _inquirer)['prompt']>() };
  const log = jest.fn<LogService>();

  beforeEach(() => {
    exec.mockReset();
    lock.take.mockReset();
    inquirer.prompt.mockReset();
    lock.release.mockReset();
    log.mockReset();
  });

  test('should work', async () => {
    exec.mockImplementationOnce((_, cb) => cb(null, 'Wayne Campbell'));
    exec.mockImplementationOnce((_, cb) => cb(null, 'wayne@warner.com'));
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      authorName: 'Wayne Campbell',
      authorEmail: 'wayne@warner.com',
    });
    lock.release.mockResolvedValueOnce(undefined);

    const author = await initAuthor({
      inquirer: inquirer as unknown as typeof _inquirer,
      exec,
      lock,
      log,
    });

    expect({
      author,
      inquirerPromptCalls: inquirer.prompt.mock.calls,
      lockTakeCalls: lock.take.mock.calls,
      lockReleaseCalls: lock.release.mock.calls,
      execCalls: exec.mock.calls,
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
    }).toMatchSnapshot();
  });

  test('should handle git failures', async () => {
    exec.mockImplementationOnce((_, cb) => cb(new Error('E_GIT_ERROR')));
    exec.mockImplementationOnce((_, cb) => cb(new Error('E_GIT_ERROR')));
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      authorName: 'Wayne Campbell',
      authorEmail: 'wayne@warner.com',
    });
    lock.release.mockResolvedValueOnce(undefined);

    const author = await initAuthor({
      inquirer: inquirer as unknown as typeof _inquirer,
      exec,
      lock,
      log,
    });

    expect({
      author,
      inquirerPromptCalls: inquirer.prompt.mock.calls,
      lockTakeCalls: lock.take.mock.calls,
      lockReleaseCalls: lock.release.mock.calls,
      execCalls: exec.mock.calls,
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
    }).toMatchSnapshot();
  });

  test('should release the lock on failure', async () => {
    exec.mockImplementationOnce((_, cb) => cb(null, 'Wayne Campbell'));
    exec.mockImplementationOnce((_, cb) => cb(null, 'wayne@warner.com'));
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockRejectedValueOnce(new Error('E_PROMPT_ERROR'));
    lock.release.mockResolvedValueOnce(undefined);

    try {
      await initAuthor({
        inquirer: inquirer as unknown as typeof _inquirer,
        exec,
        lock,
        log,
      });
    } catch (err) {
      expect({
        errorCode: (err as YError).code,
        errorParams: (err as YError).params,
        inquirerPromptCalls: inquirer.prompt.mock.calls,
        lockTakeCalls: lock.take.mock.calls,
        lockReleaseCalls: lock.release.mock.calls,
        execCalls: exec.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });
});
