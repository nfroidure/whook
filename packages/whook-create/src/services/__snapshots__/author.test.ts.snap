// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initAuthor should handle git failures 1`] = `
{
  "author": {
    "email": "wayne@warner.com",
    "name": "Wayne Campbell",
  },
  "execCalls": [
    [
      "git config --get user.name",
      [Function],
    ],
    [
      "git config --get user.email",
      [Function],
    ],
  ],
  "inquirerPromptCalls": [
    [
      [
        {
          "default": undefined,
          "message": "What's your name?",
          "name": "authorName",
          "type": "input",
        },
        {
          "default": undefined,
          "message": "Your email?",
          "name": "authorEmail",
          "type": "input",
        },
      ],
    ],
  ],
  "lockReleaseCalls": [
    [
      "cli:input",
    ],
  ],
  "lockTakeCalls": [
    [
      "cli:input",
    ],
  ],
  "logCalls": [
    [
      "debug",
      "✍️ - Initializing author...",
    ],
    [
      "debug",
      "STDERR:
",
      "",
    ],
    [
      "debug",
      "STDERR:
",
      "",
    ],
    [
      "debug",
      "Could not get author from Git",
    ],
  ],
}
`;

exports[`initAuthor should release the lock on failure 1`] = `
{
  "errorCode": "E_PROMPT_ERROR",
  "errorParams": [],
  "execCalls": [
    [
      "git config --get user.name",
      [Function],
    ],
    [
      "git config --get user.email",
      [Function],
    ],
  ],
  "inquirerPromptCalls": [
    [
      [
        {
          "default": "Wayne Campbell",
          "message": "What's your name?",
          "name": "authorName",
          "type": "input",
        },
        {
          "default": "wayne@warner.com",
          "message": "Your email?",
          "name": "authorEmail",
          "type": "input",
        },
      ],
    ],
  ],
  "lockReleaseCalls": [
    [
      "cli:input",
    ],
  ],
  "lockTakeCalls": [
    [
      "cli:input",
    ],
  ],
  "logCalls": [
    [
      "debug",
      "✍️ - Initializing author...",
    ],
  ],
}
`;

exports[`initAuthor should work 1`] = `
{
  "author": {
    "email": "wayne@warner.com",
    "name": "Wayne Campbell",
  },
  "execCalls": [
    [
      "git config --get user.name",
      [Function],
    ],
    [
      "git config --get user.email",
      [Function],
    ],
  ],
  "inquirerPromptCalls": [
    [
      [
        {
          "default": "Wayne Campbell",
          "message": "What's your name?",
          "name": "authorName",
          "type": "input",
        },
        {
          "default": "wayne@warner.com",
          "message": "Your email?",
          "name": "authorEmail",
          "type": "input",
        },
      ],
    ],
  ],
  "lockReleaseCalls": [
    [
      "cli:input",
    ],
  ],
  "lockTakeCalls": [
    [
      "cli:input",
    ],
  ],
  "logCalls": [
    [
      "debug",
      "✍️ - Initializing author...",
    ],
  ],
}
`;
